// @stylexjs/eslint-plugin/lib/stylex-valid-shorthands.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _splitShorthands = require("./utils/splitShorthands.js");
var _getSourceCode = _interopRequireDefault(require("./utils/getSourceCode"));
var _createImportTracker = _interopRequireDefault(require("./utils/createImportTracker"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const legacyNameMapping = {
  marginStart: 'marginInlineStart',
  marginEnd: 'marginInlineEnd',
  marginHorizontal: 'marginInline',
  marginVertical: 'marginBlock',
  paddingStart: 'paddingInlineStart',
  paddingEnd: 'paddingInlineEnd',
  paddingHorizontal: 'paddingInline',
  paddingVertical: 'paddingBlock'
};
const shorthandAliases = {
  background: (0, _splitShorthands.createSpecificTransformer)('background'),
  font: (0, _splitShorthands.createSpecificTransformer)('font'),
  borderColor: (0, _splitShorthands.createSpecificTransformer)('border-color'),
  borderWidth: (0, _splitShorthands.createSpecificTransformer)('border-width'),
  borderStyle: (0, _splitShorthands.createSpecificTransformer)('border-style'),
  borderTop: (0, _splitShorthands.createSpecificTransformer)('border-top'),
  borderRight: (0, _splitShorthands.createSpecificTransformer)('border-right'),
  borderBottom: (0, _splitShorthands.createSpecificTransformer)('border-bottom'),
  borderLeft: (0, _splitShorthands.createSpecificTransformer)('border-left'),
  borderRadius: (0, _splitShorthands.createSpecificTransformer)('border-radius'),
  cornerShape: (0, _splitShorthands.createSpecificTransformer)('corner-shape'),
  outline: (0, _splitShorthands.createSpecificTransformer)('outline'),
  margin: (0, _splitShorthands.createDirectionalTransformer)('margin', 'Block', 'Inline'),
  padding: (0, _splitShorthands.createDirectionalTransformer)('padding', 'Block', 'Inline'),
  marginBlock: (0, _splitShorthands.createBlockInlineTransformer)('margin', 'Block'),
  marginInline: (0, _splitShorthands.createBlockInlineTransformer)('margin', 'Inline'),
  paddingBlock: (0, _splitShorthands.createBlockInlineTransformer)('padding', 'Block'),
  paddingInline: (0, _splitShorthands.createBlockInlineTransformer)('padding', 'Inline')
};
const stylexValidShorthands = {
  meta: {
    type: 'error',
    docs: {
      description: 'Require shorthand properties to be split into individual properties',
      recommended: false,
      url: 'https://github.com/facebook/stylex/tree/main/packages/@stylexjs/eslint-plugin'
    },
    fixable: 'code',
    schema: [{
      type: 'object',
      properties: {
        validImports: {
          type: 'array',
          items: {
            oneOf: [{
              type: 'string'
            }, {
              type: 'object',
              properties: {
                from: {
                  type: 'string'
                },
                as: {
                  type: 'string'
                }
              }
            }]
          },
          default: ['stylex', '@stylexjs/stylex']
        },
        allowImportant: {
          type: 'boolean',
          default: false
        },
        preferInline: {
          type: 'boolean',
          default: false
        }
      },
      additionalProperties: false
    }]
  },
  create(context) {
    const options = context.options[0] || {};
    const {
      validImports: importsToLookFor = ['stylex', '@stylexjs/stylex']
    } = options;
    const allowImportant = options.allowImportant || false;
    const preferInline = options.preferInline || false;
    const importTracker = (0, _createImportTracker.default)(importsToLookFor);
    function isStylexCreateCallee(node) {
      return node.type === 'MemberExpression' && node.object.type === 'Identifier' && importTracker.isStylexDefaultImport(node.object.name) && node.property.type === 'Identifier' && node.property.name === 'create' || node.type === 'Identifier' && importTracker.isStylexNamedImport('create', node.name);
    }
    function validateObject(obj) {
      for (const prop of obj.properties) {
        if (prop.type === 'SpreadElement') {
          continue;
        }
        if (prop.value.type === 'ObjectExpression') {
          validateObject(prop.value);
        } else {
          validateProperty(prop);
        }
      }
    }
    function validateProperty(property) {
      if (property.computed) {
        return;
      }
      let key;
      if (property.key.type === 'Identifier') {
        key = property.key.name;
      } else if (property.key.type === 'Literal') {
        key = property.key.value;
      }
      if (typeof key === 'string' && legacyNameMapping[key] != null) {
        context.report({
          node: property,
          message: `Use "${legacyNameMapping[key]}" instead of legacy formats like "${key}" to adhere to logical property naming.`,
          fix: fixer => {
            return fixer.replaceText(property.key, legacyNameMapping[key]);
          }
        });
      }
      if (typeof key !== 'string') {
        return;
      }
      const shorthandAliasesForKey = shorthandAliases[key];
      if (typeof key !== 'string' || property.value.value === null || shorthandAliasesForKey == null) {
        return;
      }
      const v = property.value.value;
      if (typeof v !== 'string' && typeof v !== 'number') {
        return;
      }
      const newValues = shorthandAliasesForKey(v, allowImportant, preferInline);
      const isUnfixableError = newValues.length === 1 && newValues[0]?.[1] === _splitShorthands.CANNOT_FIX;
      if (!newValues || (newValues.length === 1 && newValues[0][1] === property.value.value || newValues[0][1] === property.value?.value?.toString() || newValues[0][1] === parseInt(property.value?.value, 10)) && !isUnfixableError) {
        return;
      }
      context.report({
        node: property,
        message: `Property shorthands using multiple values like "${key}: ${String(property.value.value)}" are not supported in StyleX. Separate into individual properties.`,
        data: {
          property: key
        },
        fix: !isUnfixableError ? fixer => {
          const sourceCode = (0, _getSourceCode.default)(context);
          const startNodeIndentation = getNodeIndentation(sourceCode, property);
          const newLineAndIndent = `\n${startNodeIndentation}`;
          const newPropertiesText = newValues.map(([key, value], index) => `${index > 0 ? newLineAndIndent : ''}${key}: ${typeof value === 'string' ? `'${value}'` : value}`).join(',');
          return fixer.replaceText(property, newPropertiesText);
        } : null
      });
    }
    return {
      ImportDeclaration: importTracker.ImportDeclaration,
      CallExpression(node) {
        const isStyleXCall = isStylexCreateCallee(node.callee);
        if (!isStyleXCall) {
          return;
        }
        const namespacesObj = node.arguments[0];
        if (namespacesObj.type !== 'ObjectExpression') {
          return;
        }
        for (const namespaceProp of namespacesObj.properties) {
          if (namespaceProp.type !== 'Property') {
            continue;
          }
          if (namespaceProp.value.type === 'ObjectExpression') {
            validateObject(namespaceProp.value);
          }
        }
      },
      'Program:exit'() {
        importTracker.clear();
      }
    };
  }
};
function isSameLine(aNode, bNode) {
  return Boolean(aNode.loc && bNode.loc && aNode.loc?.start.line === bNode.loc?.start.line);
}
function getNodeIndentation(sourceCode, node) {
  const tokenBefore = sourceCode.getTokenBefore(node, {
    includeComments: false
  });
  const isTokenBeforeSameLineAsNode = !!tokenBefore && isSameLine(tokenBefore, node);
  const sliceStart = isTokenBeforeSameLineAsNode && tokenBefore?.loc ? tokenBefore.loc.end.column : 0;
  return node?.loc ? sourceCode.lines[node.loc.start.line - 1].slice(sliceStart, node.loc.start.column) : '';
}
var _default = exports.default = stylexValidShorthands;
