// @stylexjs/eslint-plugin/lib/stylex-no-conflicting-props.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _createImportTracker = _interopRequireDefault(require("./utils/createImportTracker"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const stylexNoConflictingProps = {
  meta: {
    type: 'problem',
    docs: {
      description: 'Disallow using `className` or `style` props on elements that spread `stylex.props()`',
      category: 'Best Practices',
      recommended: true
    },
    schema: [{
      type: 'object',
      properties: {
        validImports: {
          type: 'array',
          items: {
            oneOf: [{
              type: 'string'
            }, {
              type: 'object',
              properties: {
                from: {
                  type: 'string'
                },
                as: {
                  type: 'string'
                }
              }
            }]
          },
          default: ['stylex', '@stylexjs/stylex']
        }
      },
      additionalProperties: false
    }]
  },
  create(context) {
    const {
      validImports: importsToLookFor = ['stylex', '@stylexjs/stylex']
    } = context.options[0] || {};
    const importTracker = (0, _createImportTracker.default)(importsToLookFor);
    function isStylexPropsCallee(node) {
      return node.type === 'MemberExpression' && node.object.type === 'Identifier' && importTracker.isStylexDefaultImport(node.object.name) && node.property.type === 'Identifier' && node.property.name === 'props' || node.type === 'Identifier' && importTracker.isStylexNamedImport('props', node.name);
    }
    return {
      ImportDeclaration: importTracker.ImportDeclaration,
      JSXOpeningElement(node) {
        const hasStylexPropsSpread = node.attributes.some(attr => attr.type === 'JSXSpreadAttribute' && attr.argument.type === 'CallExpression' && isStylexPropsCallee(attr.argument.callee));
        if (!hasStylexPropsSpread) {
          return;
        }
        for (const attr of node.attributes) {
          if (attr.type === 'JSXAttribute' && attr.name.type === 'JSXIdentifier' && (attr.name.name === 'className' || attr.name.name === 'style')) {
            context.report({
              node: attr,
              message: 'The `{{propName}}` prop should not be used when spreading `stylex.props()` to avoid conflicts.',
              data: {
                propName: attr.name.name
              }
            });
          } else if (attr.type === 'JSXSpreadAttribute' && attr.argument.type === 'ObjectExpression') {
            for (const prop of attr.argument.properties) {
              if (prop.type === 'Property' && !prop.computed && prop.key.type === 'Identifier' && (prop.key.name === 'className' || prop.key.name === 'style')) {
                context.report({
                  node: prop,
                  message: 'The `{{propName}}` prop should not be used when spreading `stylex.props()` to avoid conflicts.',
                  data: {
                    propName: prop.key.name
                  }
                });
              }
            }
          }
        }
      },
      'Program:exit'() {
        importTracker.clear();
      }
    };
  }
};
var _default = exports.default = stylexNoConflictingProps;
