// @stylexjs/eslint-plugin/lib/stylex-valid-styles.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _getDistance = _interopRequireDefault(require("./utils/getDistance"));
var _isWhiteSpaceOrEmpty = _interopRequireDefault(require("./utils/isWhiteSpaceOrEmpty"));
var _micromatch = _interopRequireDefault(require("micromatch"));
var _makeLiteralRule = _interopRequireDefault(require("./rules/makeLiteralRule"));
var _isString = _interopRequireDefault(require("./rules/isString"));
var _makeUnionRule = _interopRequireDefault(require("./rules/makeUnionRule"));
var _isNumber = _interopRequireDefault(require("./rules/isNumber"));
var _isAnimationName = _interopRequireDefault(require("./rules/isAnimationName"));
var _isPositionTryFallbacks = _interopRequireDefault(require("./rules/isPositionTryFallbacks"));
var _isStylexResolvedVarsToken = _interopRequireDefault(require("./rules/isStylexResolvedVarsToken"));
var _isCSSVariable = _interopRequireDefault(require("./rules/isCSSVariable"));
var _evaluate = _interopRequireDefault(require("./utils/evaluate"));
var _resolveKey = _interopRequireDefault(require("./utils/resolveKey"));
var _cssProperties = require("./reference/cssProperties");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const showError = message => () => ({
  message
});
const stylexValidStyles = {
  meta: {
    type: 'problem',
    hasSuggestions: true,
    fixable: 'code',
    docs: {
      descriptions: 'Enforce that you create valid stylex styles',
      category: 'Possible Errors',
      recommended: true
    },
    schema: [{
      type: 'object',
      properties: {
        validImports: {
          type: 'array',
          items: {
            oneOf: [{
              type: 'string'
            }, {
              type: 'object',
              properties: {
                from: {
                  type: 'string'
                },
                as: {
                  type: 'string'
                }
              }
            }]
          },
          default: ['stylex', '@stylexjs/stylex']
        },
        allowRawCSSVars: {
          type: 'boolean',
          default: true
        },
        allowOuterPseudoAndMedia: {
          type: 'boolean',
          default: false
        },
        banPropsForLegacy: {
          type: 'boolean',
          default: false
        },
        propLimits: {
          type: 'object',
          additionalProperties: {
            type: 'object',
            properties: {
              limit: {
                oneOf: [{
                  type: 'null'
                }, {
                  type: 'string'
                }, {
                  type: 'number'
                }, {
                  type: 'array',
                  items: {
                    oneOf: [{
                      type: 'null'
                    }, {
                      type: 'string'
                    }, {
                      type: 'number'
                    }]
                  }
                }]
              },
              reason: {
                type: 'string'
              }
            }
          }
        }
      }
    }]
  },
  create(context) {
    const variables = new Map();
    const dynamicStyleVariables = new Set();
    const options = context.options[0] || {};
    const themeFileExtension = options.themeFileExtension || '.stylex';
    const legacyReason = 'This property is not supported in legacy StyleX resolution.';
    const legacyProps = {
      'grid*': {
        limit: null,
        reason: legacyReason
      },
      'mask+([a-zA-Z])': {
        limit: null,
        reason: legacyReason
      },
      blockOverflow: {
        limit: null,
        reason: legacyReason
      },
      inlineOverflow: {
        limit: null,
        reason: legacyReason
      },
      transitionProperty: {
        limit: ['opacity', 'transform', 'opacity, transform', 'none'],
        reason: legacyReason
      }
    };
    const {
      validImports: importsToLookFor = ['stylex', '@stylexjs/stylex'],
      allowRawCSSVars = true,
      allowOuterPseudoAndMedia,
      banPropsForLegacy = false,
      propLimits = {}
    } = context.options[0] || {};
    function isValidStylexResolvedVarsFileExtension(filename, themeFileExtension) {
      const baseExtensions = [themeFileExtension, `${themeFileExtension}.const`, '.transformed'];
      const extensions = ['.js', '.ts', '.tsx', '.jsx', '.mjs', '.cjs'];
      return ['', ...extensions].some(ext => baseExtensions.some(base => filename.endsWith(`${base}${ext}`)));
    }
    const stylexResolvedVarsTokenImports = new Set();
    const styleXDefaultImports = new Set();
    const styleXCreateImports = new Set();
    const styleXKeyframesImports = new Set();
    const styleXPositionTryImports = new Set();
    const styleXWhenImports = new Set();
    const overrides = {
      ...(banPropsForLegacy ? legacyProps : {}),
      ...propLimits
    };
    const CSSPropertiesWithOverrides = {
      ..._cssProperties.CSSProperties,
      animationName: (0, _makeUnionRule.default)((0, _makeLiteralRule.default)('none'), (0, _isAnimationName.default)(styleXDefaultImports, styleXKeyframesImports), _cssProperties.all),
      positionTryFallbacks: (0, _makeUnionRule.default)((0, _makeLiteralRule.default)('none'), _isCSSVariable.default, (0, _isPositionTryFallbacks.default)(styleXDefaultImports, styleXPositionTryImports), _cssProperties.all)
    };
    for (const overrideKey in overrides) {
      const {
        limit,
        reason
      } = overrides[overrideKey];
      const overrideValue = limit === null ? showError(reason) : limit === '*' ? (0, _makeUnionRule.default)(_isString.default, _isNumber.default, _cssProperties.all) : limit === 'string' ? (0, _makeUnionRule.default)(_isString.default, _cssProperties.all) : limit === 'number' ? (0, _makeUnionRule.default)(_isNumber.default, _cssProperties.all) : typeof limit === 'string' || typeof limit === 'number' ? (0, _makeUnionRule.default)(limit, _cssProperties.all) : Array.isArray(limit) ? (0, _makeUnionRule.default)(...limit.map(l => {
        if (l === '*') {
          return (0, _makeUnionRule.default)(_isString.default, _isNumber.default);
        }
        if (l === 'string') {
          return _isString.default;
        }
        if (l === 'number') {
          return _isNumber.default;
        }
        return l;
      }), _cssProperties.all) : undefined;
      if (overrideValue === undefined) {
        continue;
      }
      if (overrideKey.includes('*') || overrideKey.includes('+')) {
        for (const key in CSSPropertiesWithOverrides) {
          if (_micromatch.default.isMatch(key, overrideKey)) {
            CSSPropertiesWithOverrides[key] = overrideValue;
          }
        }
      } else {
        CSSPropertiesWithOverrides[overrideKey] = overrideValue;
      }
    }
    function isStylexCallee(node) {
      return node.type === 'MemberExpression' && node.object.type === 'Identifier' && styleXDefaultImports.has(node.object.name) && node.property.type === 'Identifier' && node.property.name === 'create' || node.type === 'Identifier' && styleXCreateImports.has(node.name);
    }
    function isStylexDeclaration(node) {
      return node && node.type === 'CallExpression' && isStylexCallee(node.callee) && node.arguments.length === 1;
    }
    function validateStyleValue(valueNode, varsWithFnArgs, style, styleKey, propertyKey, ruleChecker) {
      if (valueNode.type === 'ConditionalExpression') {
        const trueCheck = validateStyleValue(valueNode.consequent, varsWithFnArgs, style, styleKey, propertyKey, ruleChecker);
        if (trueCheck != null) {
          return trueCheck;
        }
        const falseCheck = validateStyleValue(valueNode.alternate, varsWithFnArgs, style, styleKey, propertyKey, ruleChecker);
        if (falseCheck != null) {
          return falseCheck;
        }
        return null;
      }
      if (valueNode.type === 'LogicalExpression' && ['||', '??'].includes(valueNode.operator)) {
        const leftCheck = validateStyleValue(valueNode.left, varsWithFnArgs, style, styleKey, propertyKey, ruleChecker);
        if (leftCheck != null) {
          return leftCheck;
        }
        const rightCheck = validateStyleValue(valueNode.right, varsWithFnArgs, style, styleKey, propertyKey, ruleChecker);
        if (rightCheck != null) {
          return rightCheck;
        }
        return null;
      }
      if ((propertyKey === 'float' || propertyKey === 'clear') && valueNode.type === 'Literal' && typeof valueNode.value === 'string' && (valueNode.value === 'start' || valueNode.value === 'end')) {
        const replacement = valueNode.value === 'start' ? 'inline-start' : 'inline-end';
        return {
          node: valueNode,
          loc: valueNode.loc,
          message: `The value "${valueNode.value}" is not a standard CSS value for "${propertyKey}". Did you mean "${replacement}"?`,
          fix: fixer => fixer.replaceText(valueNode, `'${replacement}'`),
          suggest: [{
            desc: `Replace "${valueNode.value}" with "${replacement}"?`,
            fix: fixer => fixer.replaceText(valueNode, `'${replacement}'`)
          }],
          isSpecialCase: true
        };
      }
      const check = ruleChecker(valueNode, varsWithFnArgs, style, context);
      if (check != null) {
        return check;
      }
      if (valueNode.type === 'Literal' && typeof valueNode.value === 'string' && (0, _isWhiteSpaceOrEmpty.default)(valueNode.value) && styleKey.name !== 'content') {
        return {
          node: valueNode,
          loc: valueNode.loc,
          message: 'The empty string is not allowed by Stylex.',
          isSpecialCase: true
        };
      }
    }
    function checkStyleProperty(style, level, propName, outerIsPseudoElement) {
      if (style.type === 'Property') {
        if (style.value.type === 'ObjectExpression') {
          const styleValue = style.value;
          if (level > 0 && propName == null && !(outerIsPseudoElement && level === 1)) {
            return context.report({
              node: style.value,
              loc: style.value.loc,
              message: 'You cannot nest styles more than one level deep'
            });
          }
          const key = style.key;
          if (key.type === 'PrivateIdentifier') {
            return context.report({
              node: key,
              loc: key.loc,
              message: 'Private properties are not allowed in stylex'
            });
          }
          const keyName = key.type === 'Literal' ? key.value : key.type === 'Identifier' ? !style.computed ? key.name : (0, _resolveKey.default)(key, variables) : null;
          if ((0, _isStylexResolvedVarsToken.default)(key, stylexResolvedVarsTokenImports)) {
            return undefined;
          }
          if (typeof keyName !== 'string' || key.type !== 'Literal' && key.type !== 'Identifier') {
            return context.report({
              node: key,
              loc: key.loc,
              message: 'Keys must be strings'
            });
          }
          if (keyName.startsWith('@') || keyName.startsWith(':')) {
            if (level === 0) {
              const ruleCheck = (allowOuterPseudoAndMedia ? _cssProperties.allModifiers : _cssProperties.pseudoElements)(key, variables);
              if (ruleCheck !== undefined) {
                if (keyName.startsWith('::')) {
                  return context.report({
                    node: style.value,
                    loc: style.value.loc,
                    message: `Unknown pseudo element "${keyName}"`
                  });
                }
                return context.report({
                  node: style.value,
                  loc: style.value.loc,
                  message: allowOuterPseudoAndMedia ? 'Nested styles can only be used for the pseudo selectors in the stylex allowlist and for @media queries' : 'Pseudo Classes, Media Queries and other At Rules should be nested as conditions within style properties. Only Pseudo Elements (::after) are allowed at the top-level'
                });
              }
            } else {
              const ruleCheck = (0, _cssProperties.pseudoClassesAndAtRules)(key, variables);
              if (ruleCheck !== undefined) {
                return context.report({
                  node: style.value,
                  loc: style.value.loc,
                  message: 'Invalid Pseudo class or At Rule used for conditional style value'
                });
              }
            }
          }
          return styleValue.properties.forEach(prop => checkStyleProperty(prop, level + 1, propName ?? (keyName.startsWith('@') || keyName.startsWith(':') || keyName === 'default' ? null : keyName), outerIsPseudoElement || keyName.startsWith('::')));
        }
        let styleKey = style.key;
        if (styleKey.type === 'PrivateIdentifier') {
          return context.report({
            node: styleKey,
            loc: styleKey.loc,
            message: 'Private properties are not allowed in stylex'
          });
        }
        if ((0, _isStylexResolvedVarsToken.default)(styleKey, stylexResolvedVarsTokenImports)) {
          return undefined;
        }
        let isStylexWhenCall = false;
        if (style.computed && styleKey.type !== 'Literal') {
          if (styleKey.type === 'CallExpression' && styleKey.callee.type === 'MemberExpression') {
            const calleeObject = styleKey.callee.object;
            const calleeProperty = styleKey.callee.property;
            isStylexWhenCall = calleeObject.type === 'MemberExpression' && calleeObject.object.type === 'Identifier' && styleXDefaultImports.has(calleeObject.object.name) && calleeObject.property.type === 'Identifier' && calleeObject.property.name === 'when' && calleeProperty.type === 'Identifier' || calleeObject.type === 'Identifier' && styleXWhenImports.has(calleeObject.name) && calleeProperty.type === 'Identifier';
            if (!isStylexWhenCall) {
              const val = (0, _evaluate.default)(styleKey, variables);
              if (val == null) {
                return context.report({
                  node: style.key,
                  loc: style.key.loc,
                  message: 'Computed key cannot be resolved.'
                });
              } else if (val === 'ARG') {
                return context.report({
                  node: style.key,
                  loc: style.key.loc,
                  message: 'Computed key cannot depend on function argument'
                });
              } else {
                styleKey = val;
              }
            }
          }
        }
        if (styleKey.type !== 'Literal' && styleKey.type !== 'Identifier' && !isStylexWhenCall) {
          return context.report({
            node: styleKey,
            loc: styleKey.loc,
            message: 'All keys in a stylex object must be static literal values.'
          });
        }
        const key = propName ?? (styleKey.type === 'Identifier' ? styleKey.name : styleKey.value);
        if (typeof key !== 'string') {
          return context.report({
            node: styleKey,
            loc: styleKey.loc,
            message: 'All keys in a stylex object must be static literal string values.'
          });
        }
        if (_cssProperties.CSSPropertyReplacements[key] != null) {
          const propCheck = _cssProperties.CSSPropertyReplacements[key];
          const val = style;
          const check = propCheck(style.value, variables, style);
          if (check != null) {
            const {
              message,
              suggest
            } = check;
            const diagnostic = {
              node: style,
              loc: style.loc,
              message,
              suggest: suggest != null ? [suggest] : undefined
            };
            return context.report(diagnostic);
          }
        }
        const ruleChecker = CSSPropertiesWithOverrides[key];
        if (ruleChecker == null) {
          if (allowRawCSSVars && _micromatch.default.isMatch(key, '--*')) {
            return;
          }
          const closestKey = _cssProperties.CSSPropertyKeys.find(cssProp => {
            const distance = (0, _getDistance.default)(key, cssProp, 2);
            return distance <= 2;
          });
          return context.report({
            node: style.key,
            loc: style.key.loc,
            message: 'This is not a key that is allowed by stylex',
            suggest: closestKey != null ? [{
              desc: `Did you mean "${closestKey}"?`,
              fix: fixer => {
                if (style.key.type === 'Identifier') {
                  return fixer.replaceText(style.key, closestKey);
                } else if (style.key.type === 'Literal' && (typeof style.key.value === 'string' || typeof style.key.value === 'number' || typeof style.key.value === 'boolean' || style.key.value == null)) {
                  const styleKey = style.key;
                  const raw = style.key.raw;
                  if (raw != null) {
                    const quoteType = raw.substr(0, 1);
                    return fixer.replaceText(styleKey, `${quoteType}${closestKey}${quoteType}`);
                  }
                }
                return null;
              }
            }] : undefined
          });
        }
        if (typeof ruleChecker !== 'function') {
          throw new TypeError(`CSSProperties[${key}] is not a function`);
        }
        const isReferencingStylexDefineVarsTokens = stylexResolvedVarsTokenImports.size > 0 && (0, _isStylexResolvedVarsToken.default)(style.value, stylexResolvedVarsTokenImports);
        if (!isReferencingStylexDefineVarsTokens) {
          let varsWithFnArgs = variables;
          if (dynamicStyleVariables.size > 0) {
            varsWithFnArgs = new Map();
            for (const [key, value] of variables) {
              varsWithFnArgs.set(key, value);
            }
            for (const key of dynamicStyleVariables) {
              varsWithFnArgs.set(key, 'ARG');
            }
          }
          const check = validateStyleValue(style.value, varsWithFnArgs, style, styleKey, key, ruleChecker);
          if (check != null) {
            if (check.isSpecialCase) {
              return context.report({
                node: check.node,
                loc: check.loc,
                message: check.message,
                fix: check.fix,
                suggest: check.suggest
              });
            }
            const {
              message,
              suggest
            } = check;
            const isBackgroundBlendModeFormatError = key === 'backgroundBlendMode' && typeof message === 'string' && message.indexOf('backgroundBlendMode values must be separated by a comma and a space') !== -1;
            const finalMessage = isBackgroundBlendModeFormatError ? message.split('\n')[0] : `${key} value must be one of:\n${message}${key === 'lineHeight' ? '\nBe careful when fixing: lineHeight: 10px is not the same as lineHeight: 10' : ''}`;
            return context.report({
              node: style.value,
              loc: style.value.loc,
              message: finalMessage,
              suggest: suggest != null ? [suggest] : undefined
            });
          }
        }
      }
    }
    return {
      Program(node) {
        const vars = node.body.reduce((collection, node) => {
          if (node.type === 'VariableDeclaration') {
            collection.push(node);
          }
          return collection;
        }, []).map(constDecl => constDecl.declarations).reduce((arr, curr) => arr.concat(curr), []);
        const [requires, others] = vars.reduce((acc, decl) => {
          if (decl.init != null && decl.init.type === 'CallExpression' && decl.init.callee.type === 'Identifier' && decl.init.callee.name === 'require') {
            acc[0].push(decl);
          } else {
            acc[1].push(decl);
          }
          return acc;
        }, [[], []]);
        requires.forEach(decl => {
          if (decl.init != null && decl.init.type === 'CallExpression' && decl.init.callee.type === 'Identifier' && decl.init.callee.name === 'require' && decl.init.arguments.length === 1 && decl.init.arguments[0].type === 'Literal' && importsToLookFor.includes(decl.init.arguments[0].value)) {
            if (decl.id.type === 'Identifier') {
              styleXDefaultImports.add(decl.id.name);
            }
            if (decl.id.type === 'ObjectPattern') {
              decl.id.properties.forEach(prop => {
                if (prop.type === 'Property' && prop.key.type === 'Identifier' && prop.key.name === 'create' && !prop.computed && prop.value.type === 'Identifier') {
                  styleXCreateImports.add(prop.value.name);
                }
              });
            }
          }
        });
        others.filter(decl => decl.id.type === 'Identifier').forEach(decl => {
          const id = decl.id.type === 'Identifier' ? decl.id : null;
          const init = decl.init;
          if (id != null && init != null) {
            variables.set(id.name, init);
          }
        });
      },
      ImportDeclaration(node) {
        if (node.source.type !== 'Literal' || typeof node.source.value !== 'string') {
          return;
        }
        const sourceValue = node.source.value;
        const foundImportSource = importsToLookFor.find(importSource => {
          if (typeof importSource === 'string') {
            return importSource === sourceValue;
          }
          return importSource.from === sourceValue;
        });
        const isStylexImport = foundImportSource !== undefined;
        const isStylexResolvedVarsImport = isValidStylexResolvedVarsFileExtension(sourceValue, themeFileExtension);
        if (!(isStylexImport || isStylexResolvedVarsImport)) {
          return;
        }
        if (isStylexImport) {
          if (typeof foundImportSource === 'string') {
            node.specifiers.forEach(specifier => {
              if (specifier.type === 'ImportDefaultSpecifier' || specifier.type === 'ImportNamespaceSpecifier') {
                styleXDefaultImports.add(specifier.local.name);
              }
              if (specifier.type === 'ImportSpecifier' && specifier.imported.name === 'create') {
                styleXCreateImports.add(specifier.local.name);
              }
              if (specifier.type === 'ImportSpecifier' && specifier.imported.name === 'keyframes') {
                styleXKeyframesImports.add(specifier.local.name);
              }
              if (specifier.type === 'ImportSpecifier' && specifier.imported.name === 'positionTry') {
                styleXPositionTryImports.add(specifier.local.name);
              }
              if (specifier.type === 'ImportSpecifier' && specifier.imported.name === 'when') {
                styleXWhenImports.add(specifier.local.name);
              }
            });
          }
          if (typeof foundImportSource === 'object') {
            node.specifiers.forEach(specifier => {
              if (specifier.type === 'ImportSpecifier') {
                if (specifier.imported.name === foundImportSource.as) {
                  styleXDefaultImports.add(specifier.local.name);
                }
              }
            });
          }
        }
        if (isStylexResolvedVarsImport) {
          node.specifiers.forEach(specifier => {
            if (specifier.type === 'ImportSpecifier') {
              stylexResolvedVarsTokenImports.add(specifier.local.name);
            }
          });
        }
      },
      CallExpression(node) {
        if (!isStylexDeclaration(node)) {
          return;
        }
        const namespaces = node.arguments[0];
        if (namespaces.type !== 'ObjectExpression') {
          return context.report({
            node: namespaces,
            loc: namespaces.loc,
            message: 'Styles must be represented as JavaScript objects'
          });
        }
        namespaces.properties.forEach(namespace => {
          if (namespace.type !== 'Property') {
            return context.report({
              node: namespace,
              loc: namespace.loc,
              message: 'Styles cannot be spread objects'
            });
          }
          let styles = namespace.value;
          if (styles.type !== 'ObjectExpression') {
            if (styles.type === 'ArrowFunctionExpression' && (styles.body.type === 'ObjectExpression' || styles.body.type === 'TSAsExpression' && styles.body.expression.type === 'ObjectExpression')) {
              const params = styles.params;
              styles = styles.body;
              if (styles.type === 'TSAsExpression') {
                styles = styles.expression;
              }
              if (params.some(param => param.type !== 'Identifier')) {
                return params.filter(param => param.type !== 'Identifier').forEach(param => {
                  context.report({
                    node: param,
                    loc: param.loc,
                    message: 'Dynamic Styles can only accept named parameters. Destructuring, spreading or default parameters are not allowed.'
                  });
                });
              }
              params.forEach(param => {
                if (param.type === 'Identifier') {
                  dynamicStyleVariables.add(param.name);
                }
              });
            } else {
              return;
            }
          }
          styles.properties.forEach(prop => checkStyleProperty(prop, 0, null, false));
          dynamicStyleVariables.clear();
        });
      },
      'Program:exit'() {
        variables.clear();
      }
    };
  }
};
var _default = exports.default = stylexValidStyles;
