// @stylexjs/eslint-plugin/lib/stylex-no-legacy-contextual-styles.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _createImportTracker = _interopRequireDefault(require("./utils/createImportTracker"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const stylexNoLegacyContextualStyles = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Disallow legacy media query/pseudo class syntax',
      recommended: true,
      url: 'https://github.com/facebook/stylex/tree/main/packages/@stylexjs/eslint-plugin'
    },
    schema: [{
      type: 'object',
      properties: {
        validImports: {
          type: 'array',
          items: {
            oneOf: [{
              type: 'string'
            }, {
              type: 'object',
              properties: {
                from: {
                  type: 'string'
                },
                as: {
                  type: 'string'
                }
              }
            }]
          },
          default: ['stylex', '@stylexjs/stylex']
        }
      },
      additionalProperties: false
    }]
  },
  create(context) {
    const {
      validImports: importsToLookFor = ['stylex', '@stylexjs/stylex']
    } = context.options[0] || {};
    const importTracker = (0, _createImportTracker.default)(importsToLookFor);
    function isStylexCreateCallee(node) {
      return node.type === 'MemberExpression' && node.object.type === 'Identifier' && importTracker.isStylexDefaultImport(node.object.name) && node.property.type === 'Identifier' && node.property.name === 'create' || node.type === 'Identifier' && importTracker.isStylexNamedImport('create', node.name);
    }
    function isStylexCreateDeclaration(node) {
      return node && node.type === 'CallExpression' && isStylexCreateCallee(node.callee) && node.arguments.length === 1;
    }
    return {
      ImportDeclaration: importTracker.ImportDeclaration,
      CallExpression(node) {
        const firstArg = node.arguments[0];
        if (isStylexCreateDeclaration(node) && firstArg.type === 'ObjectExpression') {
          firstArg.properties.forEach(property => {
            if (property.type !== 'Property' || property.value.type !== 'ObjectExpression') {
              return;
            }
            property.value.properties.forEach(property => {
              if (property.type !== 'Property' || property.value.type !== 'ObjectExpression') {
                return;
              }
              if (property.key.type === 'Literal' && typeof property.key.value === 'string' && property.key.value.startsWith('@')) {
                context.report({
                  node: property,
                  message: 'This media query syntax is deprecated. Use the new syntax specified here: https://stylexjs.com/docs/learn/styling-ui/defining-styles/#media-queries-and-other--rules'
                });
              }
              if (property.key.type === 'Literal') {
                const literalValue = property.key.value;
                if (typeof literalValue === 'string' && literalValue.startsWith(':') && !literalValue.startsWith('::')) context.report({
                  node: property,
                  message: 'This pseudo class syntax is deprecated. Use the new syntax specified here: https://stylexjs.com/docs/learn/styling-ui/defining-styles/#pseudo-classes'
                });
              }
            });
          });
        }
      },
      'Program:exit'() {
        importTracker.clear();
      }
    };
  }
};
var _default = exports.default = stylexNoLegacyContextualStyles;
