// @stylexjs/eslint-plugin/lib/stylex-no-nonstandard-styles.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _makeLiteralRule = _interopRequireDefault(require("./rules/makeLiteralRule"));
var _makeUnionRule = _interopRequireDefault(require("./rules/makeUnionRule"));
var _isAnimationName = _interopRequireDefault(require("./rules/isAnimationName"));
var _isPositionTryFallbacks = _interopRequireDefault(require("./rules/isPositionTryFallbacks"));
var _isStylexResolvedVarsToken = _interopRequireDefault(require("./rules/isStylexResolvedVarsToken"));
var _isCSSVariable = _interopRequireDefault(require("./rules/isCSSVariable"));
var _evaluate = _interopRequireDefault(require("./utils/evaluate"));
var _resolveKey = _interopRequireDefault(require("./utils/resolveKey"));
var _cssProperties = require("./reference/cssProperties");
var _createImportTracker = _interopRequireDefault(require("./utils/createImportTracker"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const stylexNoNonstandardStyles = {
  meta: {
    type: 'problem',
    hasSuggestions: true,
    fixable: 'code',
    docs: {
      descriptions: 'Enforce that you create standard CSS values and properties for stylex',
      category: 'Possible Errors',
      recommended: true
    },
    schema: [{
      type: 'object',
      properties: {
        validImports: {
          type: 'array',
          items: {
            oneOf: [{
              type: 'string'
            }, {
              type: 'object',
              properties: {
                from: {
                  type: 'string'
                },
                as: {
                  type: 'string'
                }
              }
            }]
          },
          default: ['stylex', '@stylexjs/stylex']
        }
      }
    }]
  },
  create(context) {
    const {
      validImports: importsToLookFor = ['stylex', '@stylexjs/stylex']
    } = context.options[0] || {};
    const importTracker = (0, _createImportTracker.default)(importsToLookFor);
    const variables = new Map();
    const dynamicStyleVariables = new Set();
    const stylexResolvedVarsTokenImports = new Set();
    const styleXDefaultImports = new Set();
    const styleXCreateImports = new Set();
    const styleXKeyframesImports = new Set();
    const styleXPositionTryImports = new Set();
    const CSSPropertiesWithOverrides = {
      ..._cssProperties.CSSProperties,
      animationName: (0, _makeUnionRule.default)((0, _makeLiteralRule.default)('none'), (0, _isAnimationName.default)(styleXDefaultImports, styleXKeyframesImports), _cssProperties.all),
      positionTryFallbacks: (0, _makeUnionRule.default)((0, _makeLiteralRule.default)('none'), _isCSSVariable.default, (0, _isPositionTryFallbacks.default)(styleXDefaultImports, styleXPositionTryImports), _cssProperties.all)
    };
    function isStylexCreateCallee(node) {
      return node.type === 'MemberExpression' && node.object.type === 'Identifier' && importTracker.isStylexDefaultImport(node.object.name) && node.property.type === 'Identifier' && node.property.name === 'create' || node.type === 'Identifier' && importTracker.isStylexNamedImport('create', node.name);
    }
    function isStylexCreateDeclaration(node) {
      return node && node.type === 'CallExpression' && isStylexCreateCallee(node.callee) && node.arguments.length === 1;
    }
    function checkStyleProperty(style, level, propName, outerIsPseudoElement) {
      if (style.type !== 'Property') {
        return;
      }
      if (style.value.type === 'ObjectExpression') {
        const styleValue = style.value;
        if (level > 0 && propName == null && !(outerIsPseudoElement && level === 1)) {
          return;
        }
        const key = style.key;
        if (key.type === 'PrivateIdentifier') {
          return;
        }
        const keyName = key.type === 'Literal' ? key.value : key.type === 'Identifier' ? !style.computed ? key.name : (0, _resolveKey.default)(key, variables) : null;
        if ((0, _isStylexResolvedVarsToken.default)(key, stylexResolvedVarsTokenImports)) {
          return;
        }
        if (typeof keyName !== 'string' || key.type !== 'Literal' && key.type !== 'Identifier') {
          return;
        }
        return styleValue.properties.forEach(prop => {
          const nestedKeyName = propName ?? (keyName.startsWith('@') || keyName.startsWith(':') || keyName === 'default' ? null : keyName);
          if (nestedKeyName == null) {
            return undefined;
          }
          return checkStyleProperty(prop, level + 1, nestedKeyName, outerIsPseudoElement || keyName.startsWith('::'));
        });
      }
      let styleKey = style.key;
      if (styleKey.type === 'PrivateIdentifier' || (0, _isStylexResolvedVarsToken.default)(styleKey, stylexResolvedVarsTokenImports)) {
        return;
      }
      if (style.computed && styleKey.type !== 'Literal') {
        const val = (0, _evaluate.default)(styleKey, variables);
        if (val != null && val !== 'ARG') {
          styleKey = val;
        }
      }
      const key = propName ?? (styleKey.type === 'Identifier' ? styleKey.name : styleKey.value);
      if (typeof key !== 'string') {
        return;
      }
      const ruleChecker = CSSPropertiesWithOverrides[key];
      if (ruleChecker == null) {
        const replacementKey = style.key.type === 'Identifier' && _cssProperties.convertToStandardProperties[style.key.name] ? _cssProperties.convertToStandardProperties[style.key.name] : style.key.type === 'Literal' && typeof style.key.value === 'string' && _cssProperties.convertToStandardProperties[style.key.value] ? _cssProperties.convertToStandardProperties[style.key.value] : null;
        let originalKey = '';
        if (style.key.type === 'Identifier') {
          originalKey = style.key.name;
        } else if (style.key.type === 'Literal' && typeof style.key.value === 'string') {
          originalKey = style.key.value;
        }
        if (replacementKey && (style.key.type === 'Identifier' || style.key.type === 'Literal')) {
          return context.report({
            node: style.key,
            loc: style.key.loc,
            message: `The key "${originalKey}" is not a standard CSS property. Did you mean "${replacementKey}"?`,
            fix: fixer => {
              return fixer.replaceText(style.key, replacementKey);
            }
          });
        }
        return;
      }
      if (typeof ruleChecker !== 'function') {
        throw new TypeError(`CSSProperties[${key}] is not a function`);
      }
      const isReferencingStylexDefineVarsTokens = stylexResolvedVarsTokenImports.size > 0 && (0, _isStylexResolvedVarsToken.default)(style.value, stylexResolvedVarsTokenImports);
      if (!isReferencingStylexDefineVarsTokens) {
        let varsWithFnArgs = variables;
        if (dynamicStyleVariables.size > 0) {
          varsWithFnArgs = new Map();
          for (const [key, value] of variables) {
            varsWithFnArgs.set(key, value);
          }
          for (const key of dynamicStyleVariables) {
            varsWithFnArgs.set(key, 'ARG');
          }
        }
        if ((key === 'float' || key === 'clear') && style.value.type === 'Literal' && typeof style.value.value === 'string' && (style.value.value === 'start' || style.value.value === 'end')) {
          const replacement = style.value.value === 'start' ? 'inline-start' : 'inline-end';
          return context.report({
            node: style.value,
            loc: style.value.loc,
            message: `The value "${style.value.value}" is not a standard CSS value for "${key}". Did you mean "${replacement}"?`,
            fix: fixer => fixer.replaceText(style.value, `'${replacement}'`),
            suggest: [{
              desc: `Replace "${style.value.value}" with "${replacement}"?`,
              fix: fixer => fixer.replaceText(style.value, `'${replacement}'`)
            }]
          });
        }
      }
    }
    return {
      ImportDeclaration: importTracker.ImportDeclaration,
      Program(node) {
        const vars = node.body.reduce((collection, node) => {
          if (node.type === 'VariableDeclaration') {
            collection.push(node);
          }
          return collection;
        }, []).map(constDecl => constDecl.declarations).reduce((arr, curr) => arr.concat(curr), []);
        const [requires, others] = vars.reduce((acc, decl) => {
          if (decl.init != null && decl.init.type === 'CallExpression' && decl.init.callee.type === 'Identifier' && decl.init.callee.name === 'require') {
            acc[0].push(decl);
          } else {
            acc[1].push(decl);
          }
          return acc;
        }, [[], []]);
        requires.forEach(decl => {
          if (decl.init != null && decl.init.type === 'CallExpression' && decl.init.callee.type === 'Identifier' && decl.init.callee.name === 'require' && decl.init.arguments.length === 1 && decl.init.arguments[0].type === 'Literal' && importsToLookFor.includes(decl.init.arguments[0].value)) {
            if (decl.id.type === 'Identifier') {
              styleXDefaultImports.add(decl.id.name);
            }
            if (decl.id.type === 'ObjectPattern') {
              decl.id.properties.forEach(prop => {
                if (prop.type === 'Property' && prop.key.type === 'Identifier' && prop.key.name === 'create' && !prop.computed && prop.value.type === 'Identifier') {
                  styleXCreateImports.add(prop.value.name);
                }
              });
            }
          }
        });
        others.filter(decl => decl.id.type === 'Identifier').forEach(decl => {
          const id = decl.id.type === 'Identifier' ? decl.id : null;
          const init = decl.init;
          if (id != null && init != null) {
            variables.set(id.name, init);
          }
        });
      },
      CallExpression(node) {
        if (!isStylexCreateDeclaration(node)) {
          return;
        }
        const namespaces = node.arguments[0];
        if (namespaces.type !== 'ObjectExpression') {
          return;
        }
        namespaces.properties.forEach(property => {
          if (property.type !== 'Property') {
            return;
          }
          let styles = property.value;
          if (styles.type === 'ArrowFunctionExpression' && (styles.body.type === 'ObjectExpression' || styles.body.type === 'TSAsExpression' && styles.body.expression.type === 'ObjectExpression')) {
            const params = styles.params;
            styles = styles.type === 'TSAsExpression' ? styles.expression : styles.body;
            params.forEach(param => {
              if (param.type === 'Identifier') {
                dynamicStyleVariables.add(param.name);
              }
            });
          } else if (styles.type !== 'ObjectExpression') {
            return;
          }
          styles.properties.forEach(prop => checkStyleProperty(prop, 0, null, false));
          dynamicStyleVariables.clear();
        });
      },
      'Program:exit'() {
        importTracker.clear();
        variables.clear();
      }
    };
  }
};
var _default = exports.default = stylexNoNonstandardStyles;
