// @stylexjs/eslint-plugin/lib/stylex-no-unused.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _getSourceCode = _interopRequireDefault(require("./utils/getSourceCode"));
var _createImportTracker = _interopRequireDefault(require("./utils/createImportTracker"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function getPropertiesByName(node) {
  const properties = new Map();
  if (node == null) {
    return properties;
  }
  node.properties?.filter(property => !property.computed && !property.method).forEach(property => {
    const {
      key
    } = property;
    if (key?.type === 'Identifier') {
      properties.set(key.name, property);
    } else if (key?.type === 'Literal') {
      properties.set(key.value, property);
    }
  });
  return properties;
}
const stylexNoUnused = {
  meta: {
    fixable: 'code',
    schema: [{
      type: 'object',
      properties: {
        validImports: {
          type: 'array',
          items: {
            oneOf: [{
              type: 'string'
            }, {
              type: 'object',
              properties: {
                from: {
                  type: 'string'
                },
                as: {
                  type: 'string'
                }
              }
            }]
          },
          default: ['stylex', '@stylexjs/stylex']
        }
      },
      additionalProperties: false
    }]
  },
  create(context) {
    const {
      validImports: importsToLookFor = ['stylex', '@stylexjs/stylex']
    } = context.options[0] || {};
    const importTracker = (0, _createImportTracker.default)(importsToLookFor);
    const stylexProperties = new Map();
    function isStylexCreate(node) {
      return node.type === 'MemberExpression' && node.object.type === 'Identifier' && importTracker.isStylexDefaultImport(node.object.name) && node.property.type === 'Identifier' && node.property.name === 'create' || node.type === 'Identifier' && importTracker.isStylexNamedImport('create', node.name);
    }
    function isStylexDeclaration(node) {
      return node && node.type === 'CallExpression' && isStylexCreate(node.callee) && node.arguments.length === 1 && node.arguments[0].type === 'ObjectExpression';
    }
    function saveStylexCalls(node) {
      const id = node.id;
      const init = node.init;
      if (id && id.type === 'Identifier' && init && isStylexDeclaration(init)) {
        stylexProperties.set(id.name, getPropertiesByName(init.arguments && init.arguments?.length > 0 ? init.arguments[0] : null));
      }
    }
    function checkArguments(namespaces) {
      return function (argument) {
        if (argument) {
          if (argument.type === 'Literal') {
            namespaces.delete(argument.value);
          } else if (argument.type === 'ObjectExpression') {
            argument.properties.forEach(property => {
              if (property.key) {
                namespaces.delete(property.key.name);
              }
            });
          } else if (argument.type === 'ArrayExpression') {
            argument.elements.forEach(element => {
              namespaces.delete(element?.value);
            });
          } else if (argument.type === 'ConditionalExpression') {
            const {
              consequent,
              alternate
            } = argument;
            checkArguments(namespaces)(consequent);
            checkArguments(namespaces)(alternate);
          } else if (argument.type === 'LogicalExpression' && argument.operator === '&&') {
            checkArguments(namespaces)(argument.right);
          }
        }
      };
    }
    return {
      Program(node) {
        node.body.filter(node => node.type === 'ImportDeclaration').forEach(importTracker.ImportDeclaration);
        node.body.filter(({
          type
        }) => type === 'VariableDeclaration').map(({
          declarations
        }) => declarations && declarations.length === 1 ? declarations[0] : null).filter(Boolean).filter(({
          init
        }) => init && isStylexDeclaration(init)).forEach(saveStylexCalls);
      },
      MemberExpression(node) {
        if (node.object.type === 'Identifier' && stylexProperties.has(node.object.name)) {
          if (node.computed && node.property.type !== 'Literal') {
            stylexProperties.delete(node.object.name);
          } else if (node.property.type === 'Identifier') {
            stylexProperties.get(node.object.name)?.delete(node.property.name);
          } else if (node.property.type === 'Literal') {
            stylexProperties.get(node.object.name)?.delete(node.property.value);
          }
        }
      },
      CallExpression(node) {
        const functionName = node.callee?.name;
        if (functionName == null || !stylexProperties.has(functionName)) {
          return;
        }
        const namespaces = stylexProperties.get(functionName);
        if (namespaces == null) {
          return;
        }
        node.arguments?.forEach(checkArguments(namespaces));
      },
      ExportNamedDeclaration(node) {
        const declarations = node.declaration?.declarations;
        if (declarations?.length !== 1) {
          return;
        }
        const exportName = declarations[0].id.name;
        if (exportName == null || !stylexProperties.has(exportName)) {
          return;
        }
        stylexProperties.delete(exportName);
      },
      ExportDefaultDeclaration(node) {
        const exportName = node.declaration.name;
        if (exportName == null || !stylexProperties.has(exportName)) {
          return;
        }
        stylexProperties.delete(exportName);
      },
      ReturnStatement(node) {
        if (node.argument?.type === 'Identifier') {
          const returnName = node.argument.name;
          if (stylexProperties.has(node.argument.name)) {
            stylexProperties.delete(returnName);
          }
        }
      },
      'Program:exit'() {
        const sourceCode = (0, _getSourceCode.default)(context);
        stylexProperties.forEach((namespaces, varName) => {
          namespaces.forEach((node, namespaceName) => {
            context.report({
              node,
              message: `Unused style detected: ${varName}.${namespaceName}`,
              fix(fixer) {
                const commaOffset = sourceCode.getTokenAfter(node, {
                  includeComments: false
                })?.value === ',' ? 1 : 0;
                const left = sourceCode.getTokenBefore(node, {
                  includeComments: false
                });
                if (node.range == null || left?.range == null) {
                  return null;
                }
                return fixer.removeRange([left.range[1], node.range[1] + commaOffset]);
              }
            });
          });
        });
        stylexProperties.clear();
        importTracker.clear();
      }
    };
  }
};
var _default = exports.default = stylexNoUnused;
