// @stylexjs/eslint-plugin/lib/utils/createImportTracker.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createImportTracker;
function createImportTracker(importsToLookFor) {
  const styleXDefaultImports = new Set();
  const styleXNamedImports = new Map();
  function handleImportDeclaration(node) {
    if (node.source.type !== 'Literal' || typeof node.source.value !== 'string') {
      return;
    }
    const foundImportSource = importsToLookFor.find(importSource => {
      if (typeof importSource === 'string') {
        return importSource === node.source?.value;
      }
      return importSource.from === node.source?.value;
    });
    if (!foundImportSource) {
      return;
    }
    if (typeof foundImportSource === 'string') {
      node.specifiers.forEach(specifier => {
        if (specifier.type === 'ImportDefaultSpecifier' || specifier.type === 'ImportNamespaceSpecifier') {
          styleXDefaultImports.add(specifier.local.name);
        }
        if (specifier.type === 'ImportSpecifier') {
          const importName = specifier.imported.name;
          if (!styleXNamedImports.has(importName)) {
            styleXNamedImports.set(importName, new Set());
          }
          styleXNamedImports.get(importName)?.add(specifier.local.name);
        }
      });
    }
    if (typeof foundImportSource === 'object') {
      node.specifiers.forEach(specifier => {
        if (specifier.type === 'ImportSpecifier') {
          if (specifier.imported.name === foundImportSource.as) {
            styleXDefaultImports.add(specifier.local.name);
          }
        }
      });
    }
  }
  function isStylexDefaultImport(name) {
    return styleXDefaultImports.has(name);
  }
  function isStylexNamedImport(importName, name) {
    return styleXNamedImports.get(importName)?.has(name) ?? false;
  }
  function clear() {
    styleXDefaultImports.clear();
    styleXNamedImports.clear();
  }
  return {
    ImportDeclaration: handleImportDeclaration,
    isStylexDefaultImport,
    isStylexNamedImport,
    clear
  };
}
